mantidproject/mantid
# Only Show Runs in the Data Cache When It is Available

### Description of work

#### Summary of work
When the Data Cache is available (such as on IDAaaS), only the runs that are available in it will be shown in the investigation menu, and therefore be available for transfer and auto-processing. 

This is to account for two problems:
1. Not all of the files in the archive are available in the data cache (yet).
2. Items are added to the archive (which is what we query with the invesigation ID to get the runs list) _before_ the runs are available in the data cache. 

Making sure only the items in the data cache before showing them to auto-processing ensures that we don't just instantly fail by not finding the file. 

Fixes #36860 

#### Further detail of work

- Adds some functionality to the ISISInstrumentDataCache class to check if the cache is available and to get the list of runs available within it. 
- Makes a change to the conversion to the search results generator, so that it filters out any of the runs that aren't present in the data cache. 

### To test:
The easiest way to do this is probably to just make a package build to use on IDAaaS, but if you want to test it locally, here you go (you'll need access to the archive or at least a decent number of INTER files with the same investigation ID):

1. Boot workbench and open the ISIS Reflectometry Interface
2. Enter 1120015 and 11_3
3. Use Ctrl+Shift to select a bunch of files and transfer them to the runs table
4. Process the files and save out the workspaces from the TOF group.
5. In your file browser rename the files to INTER000[11991].nxs (replacing the run number with the correct one)
6. Pick a location on your machine for your fake data cache. 
7. Have the files in a directory named after the instrument. `INTER` in this case. 
8. In the `INTER` directory, create an `INTER_index.json` file containing the following (making the run numbers match, obviously): 
```json
{
   "11990": "11_3",
    "11991": "11_3",
    "11992": "11_3",
    "11993": "11_3",
    "11994": "11_3",
    "HIDE_11995": "11_3",
    "11996": "11_3",
    "11997": "11_3",
    "11998": "11_3",
    "HIDE_11999": "11_3",
    "12000": "11_3",
    "12001": "11_3",
    "12002": "11_3",
    "12003": "11_3",
    "12004": "11_3"
}
```
10. in the end, your structure should look like this:
<img width="681" alt="image" src="https://github.com/mantidproject/mantid/assets/47181718/c0dd7e05-7ab8-4efd-8c42-713041ab06c3">

11. Final bit of setup, set the datacache location in your `mantid.user.properties` file: `datacachesearch.directory=/path/to/FakeCache`
12. Then restart mantid to pick up the change to the properties
13. Open the Refl interface
14. Enter your run number and cycle from before
15. Click autoprocess
16. Only the files present in the index file should be moved over to the runs table
17. Edit the index file to remove the `HIDE` parts.
18. The autoprocessing should pick up the newly added runs and process them automatically. 


---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
